File: algorithms/java/113_path_sum_2/PathSum.java
class PathSum {
	public List<List<Integer>> pathSum(TreeNode root, int sum) {
        if (root == null) {
            return new ArrayList<>();
        }
        List<List<Integer>> res = new ArrayList<>();
        pathSum(res, sum, new ArrayList<>(), 0, root);
        return res;
    }
    private void pathSum(List<List<Integer>> res, int sum, List<Integer> curPath, int curSum, TreeNode root) {
        curSum += root.val;
        curPath.add(root.val);
        if (root.left == null && root.right == null && curSum == sum) {
            res.add(new ArrayList<>(curPath));
            curPath.remove(curPath.size() - 1);
            return;
        }
        if (root.left != null) {
            pathSum(res, sum, curPath, curSum, root.left);
        }
        if (root.right != null) {
            pathSum(res, sum, curPath, curSum, root.right);
        }
        curPath.remove(curPath.size() - 1);
    }
}

File: algorithms/cpp/113_path_sum2/PathSum.cpp
class PathSum {
public:
	vector<vector<int>> pathSum(TreeNode* root, int target_sum) {
        if (root == NULL) {
            return {};
        }
        vector<vector<int>> res;
        vector<int> cur;
        dfs(res, cur, root, target_sum);
        return res;
    }
    void dfs(vector<vector<int>>& res, vector<int>& cur, TreeNode* root, int target_sum) {
        cur.push_back(root->val);
        if (root->left == NULL && root->right == NULL) {
            if (target_sum == root->val) {
                res.push_back(cur);
            }
            cur.pop_back();
            return;
        }
        if (root->left) 
            dfs(res, cur, root->left, target_sum - root->val);
        if (root->right)
            dfs(res, cur, root->right, target_sum - root->val);
        cur.pop_back();
    }
};
